File: packages/mapping/src/mapper.ts
import type { ControlFrameEvent, HandControl } from "@hand-gestures/protocol";
import { DistanceCalibrator } from "./distance";
import type { SmoothedFrame, SmoothedHandFeatures } from "./expression";
import { GateLatch } from "./gateLatch";
import { PresenceHold } from "./presence";
import {
  degreeToScaleMidi,
  scaleSize,
  ScaleQuantizer,
} from "./scale";
import type { InstrumentConfig } from "./types";

const DEFAULT_VOLUME = 0.75;
const VICTORY = "Victory";
const EDGE = 0.02;
const RELEASE_LOW = 0.06;
const GATE_RESET_FRAMES = 5;

function axisValue(hand: SmoothedHandFeatures, axis: string): number {
  if (axis === "x") return hand.x_norm;
  if (axis === "z") return hand.z_norm;
  return hand.y_norm;
}

export class GestureMapper {
  private octaveShift = 0;
  private lastVolume = DEFAULT_VOLUME;
  private lastPitchMidi: number | null = null;
  private lastHandsDistance = 0;
  private rightGate: GateLatch;
  private rightPresence = new PresenceHold(8);
  private victoryHold = new PresenceHold(3);
  private distance: DistanceCalibrator;
  private pitchQuantizer: ScaleQuantizer;
  private edgeLock: "low" | "high" | null = null;
  private lastPublishedDegree: number | null = null;
  private gateClosedFrames = 0;

  constructor(private config: InstrumentConfig) {
    this.rightGate = GateLatch.fromLists(
      config.gestureRoleRight.gateOn,
      config.gestureRoleRight.gateOff
    );
    this.distance = new DistanceCalibrator(
      config.pairVolume.volumeDistMin,
      config.pairVolume.volumeDistMax
    );
    this.pitchQuantizer = new ScaleQuantizer(
      config.scaleName,
      config.scaleHysteresis
    );
  }

  get preset(): string {
    return this.config.preset;
  }

  get scaleName(): string {
    return this.config.scaleName;
  }

  applyOctaveShift(delta: number): void {
    this.octaveShift = Math.max(-2, Math.min(2, this.octaveShift + delta));
  }

  mapFrame(
    smoothed: SmoothedFrame,
    timestampMs: number,
    leftGesture: string,
    rightGesture: string,
    rightGestureStable: string
  ): ControlFrameEvent {
    const right = this.mapPitchHand(
      smoothed.right,
      rightGesture,
      rightGestureStable
    );
    const left = this.mapHand(
      smoothed.left,
      this.config.roleLeft.pitchAxis,
      leftGesture
    );

    const gateOpen = right.gate_open === true;
    const volActive = this.victoryHold.update(rightGesture === VICTORY);
    const vol = this.volumeForPair(smoothed, gateOpen, volActive, timestampMs);

    return {
      schema_version: "2",
      type: "control.frame",
      timestamp_ms: timestampMs,
      preset: this.preset,
      scale: this.scaleName,
      left,
      right,
      pair: {
        hands_distance: smoothed.hands_distance,
        volume_master: vol,
        spread: smoothed.spread,
        volume_active: volActive,
      },
    };
  }

  private volumeForPair(
    smoothed: SmoothedFrame,
    gateOpen: boolean,
    volActive: boolean,
    timestampMs: number
  ): number {
    if (!gateOpen) return 0;
    if (!volActive) return this.lastVolume;

    const dist =
      smoothed.left.presence && smoothed.right.presence
        ? smoothed.hands_distance
        : this.lastHandsDistance;

    if (smoothed.left.presence && smoothed.right.presence) {
      this.lastHandsDistance = smoothed.hands_distance;
    }

    const vol = this.distance.map(dist, true, timestampMs);
    this.lastVolume = vol;
    return vol;
  }

  private mapPitchHand(
    hand: SmoothedHandFeatures,
    gesture: string,
    gateGestureStable: string
  ): HandControl {
    const pitchAxis = this.config.roleRight.pitchAxis;
    const pitchNorm = axisValue(hand, pitchAxis);

    const gate = this.rightGate.update(
      gesture,
      hand.presence,
      gateGestureStable
    );

    let presence = hand.presence;
    if (this.preset === "scale_gate") {
      presence = gate ? true : this.rightPresence.update(hand.presence);
    }

    let pitchMidi: number | null = null;
    let degree: number | null = null;

    if (this.preset === "scale_gate") {
      if (!gate) {
        this.gateClosedFrames += 1;
        if (this.gateClosedFrames >= GATE_RESET_FRAMES) {
          this.pitchQuantizer.reset();
          this.edgeLock = null;
          this.lastPublishedDegree = null;
        }
      } else {
        this.gateClosedFrames = 0;
      }
    }

    if (this.preset === "scale_gate" && (presence || gate)) {
      if (gate) {
        degree = this.resolvePitchDegree(pitchNorm);
        this.pitchQuantizer.syncDegree(degree);
        pitchMidi = degreeToScaleMidi(degree, this.scaleName, this.octaveShift);
        this.lastPitchMidi = pitchMidi;
        this.lastPublishedDegree = degree;
      } else if (this.lastPitchMidi != null) {
        pitchMidi = this.lastPitchMidi;
        degree = this.lastPublishedDegree;
      }
    }

    return {
      presence: presence || gate,
      x: hand.spatial_x,
      y: hand.spatial_y,
      z: hand.z_norm,
      pan: 0,
      pitch_norm: pitchNorm,
      mod: 0,
      gesture: gesture !== "None" ? gesture : null,
      gate_open: gate,
      scale_degree: degree,
      spatial_x: hand.spatial_x,
      spatial_y: hand.spatial_y,
      pitch_midi: gate ? pitchMidi : null,
    };
  }

  private mapHand(
    hand: SmoothedHandFeatures,
    pitchAxis: string,
    gesture: string
  ): HandControl {
    const pitchNorm = axisValue(hand, pitchAxis);
    return {
      presence: hand.presence,
      x: hand.spatial_x,
      y: hand.spatial_y,
      z: hand.z_norm,
      pan: 0,
      pitch_norm: pitchNorm,
      mod: 0,
      gesture: gesture !== "None" ? gesture : null,
      gate_open: false,
      scale_degree: null,
      spatial_x: hand.spatial_x,
      spatial_y: hand.spatial_y,
      pitch_midi: null,
    };
  }

  private resolvePitchDegree(pitchNorm: number): number {
    const n = scaleSize(this.scaleName);
    const maxDeg = Math.max(0, n - 1);
    const releaseHigh = maxDeg > 1 ? 0.5 - EDGE : 1 - EDGE;

    if (pitchNorm <= EDGE) {
      this.edgeLock = "low";
    } else if (this.edgeLock === "low" && pitchNorm > RELEASE_LOW) {
      this.edgeLock = null;
    }

    if (pitchNorm >= 1 - EDGE) {
      this.edgeLock = "high";
    } else if (this.edgeLock === "high" && pitchNorm < releaseHigh) {
      this.edgeLock = null;
    }

    if (this.edgeLock === "low") {
      return 0;
    }
    if (this.edgeLock === "high") {
      return maxDeg;
    }

    return this.pitchQuantizer.quantize(pitchNorm);
  }

  gestureAction(gesture: string, hand: string | null): { action?: string; hand?: string } | null {
    const trigger = this.config.gestureTriggers[gesture];
    if (!trigger) return null;
    const th = (trigger.hand ?? "any").toLowerCase();
    if (th !== "any" && hand && th !== hand.toLowerCase()) return null;
    return trigger;
  }
}

File: src/mapping/mapper.py
"""Features suavizadas → ControlFrameEvent."""

from __future__ import annotations

from src.events import ControlFrameEvent, HandControl, PairControl
from src.expression.distance import DistanceCalibrator
from src.expression.engine import SmoothedFrame, SmoothedHandFeatures
from src.expression.presence import PresenceHold
from src.mapping.gate_latch import GateLatch
from src.mapping.loader import HandRole, InstrumentConfig
from src.mapping.scale import (
    ScaleQuantizer,
    degree_to_scale_midi,
    scale_size,
)

DEFAULT_VOLUME = 0.75
VICTORY_GESTURE = "Victory"
EDGE = 0.02
RELEASE_LOW = 0.06
GATE_RESET_FRAMES = 5


def _axis_value(hand: SmoothedHandFeatures, axis: str) -> float:
    if axis == "x":
        return hand.x_norm
    if axis == "z":
        return hand.z_norm
    return hand.y_norm


class GestureMapper:
    def __init__(self, config: InstrumentConfig) -> None:
        self._config = config
        self._preset = config.preset
        self._scale_name = config.scale.name
        self._octave_shift = 0
        self._last_volume = DEFAULT_VOLUME
        self._last_pitch_midi: float | None = None
        self._last_hands_distance = 0.0
        self._right_gate = GateLatch(
            config.gesture_role_right.gate_on,
            config.gesture_role_right.gate_off,
        )
        self._right_presence = PresenceHold(hold_frames=8)
        self._victory_hold = PresenceHold(hold_frames=3)
        self._distance = DistanceCalibrator(
            config.pair_volume.volume_dist_min,
            config.pair_volume.volume_dist_max,
        )
        self._pitch_quantizer = ScaleQuantizer(
            config.scale.name,
            config.scale.hysteresis,
        )
        self._edge_lock: str | None = None
        self._last_published_degree: int | None = None
        self._gate_closed_frames = 0

    @property
    def preset(self) -> str:
        return self._preset

    @property
    def scale_name(self) -> str:
        return self._scale_name

    def apply_octave_shift(self, delta: int) -> None:
        self._octave_shift = max(-2, min(2, self._octave_shift + delta))

    def map_frame(
        self,
        smoothed: SmoothedFrame,
        timestamp_ms: int,
        left_gesture: str = "None",
        right_gesture: str = "None",
        right_gesture_stable: str = "None",
    ) -> ControlFrameEvent:
        right = self._map_pitch_hand(
            smoothed.right, right_gesture, right_gesture_stable
        )
        left = self._map_hand(
            smoothed.left,
            self._config.role_left,
            left_gesture,
            pitch_hand=False,
        )

        gate_open = right.gate_open
        vol_active = self._victory_hold.update(right_gesture == VICTORY_GESTURE)
        vol = self._volume_for_pair(smoothed, gate_open, vol_active, timestamp_ms)

        return ControlFrameEvent(
            timestamp_ms=timestamp_ms,
            preset=self._preset,
            scale=self._scale_name,
            left=left,
            right=right,
            pair=PairControl(
                hands_distance=smoothed.hands_distance,
                volume_master=vol,
                spread=smoothed.spread,
                volume_active=vol_active,
            ),
        )

    def _volume_for_pair(
        self,
        smoothed: SmoothedFrame,
        gate_open: bool,
        vol_active: bool,
        timestamp_ms: int,
    ) -> float:
        if not gate_open:
            return 0.0
        if not vol_active:
            return self._last_volume

        if smoothed.left.presence and smoothed.right.presence:
            dist = smoothed.hands_distance
            self._last_hands_distance = dist
        else:
            dist = self._last_hands_distance

        vol = self._distance.map(dist, True, timestamp_ms)
        self._last_volume = vol
        return vol

    def _map_pitch_hand(
        self,
        hand: SmoothedHandFeatures,
        gesture: str,
        gate_gesture_stable: str,
    ) -> HandControl:
        role = self._config.role_right
        pitch_norm = _axis_value(hand, role.pitch_axis)

        gate = self._right_gate.update(gesture, hand.presence, gate_gesture_stable)

        presence = hand.presence
        if self._preset == "scale_gate":
            presence = True if gate else self._right_presence.update(hand.presence)

        pitch_midi: float | None = None
        degree: int | None = None

        if self._preset == "scale_gate":
            if not gate:
                self._gate_closed_frames += 1
                if self._gate_closed_frames >= GATE_RESET_FRAMES:
                    self._pitch_quantizer.reset()
                    self._edge_lock = None
                    self._last_published_degree = None
            else:
                self._gate_closed_frames = 0

        if self._preset == "scale_gate" and (presence or gate):
            if gate:
                degree = self._resolve_pitch_degree(pitch_norm)
                self._pitch_quantizer.sync_degree(degree)
                pitch_midi = degree_to_scale_midi(
                    degree, self._scale_name, self._octave_shift
                )
                self._last_pitch_midi = pitch_midi
                self._last_published_degree = degree
            elif self._last_pitch_midi is not None:
                pitch_midi = self._last_pitch_midi
                degree = self._last_published_degree

        sx = getattr(hand, "spatial_x", hand.x_norm)
        sy = getattr(hand, "spatial_y", hand.y_norm)

        return HandControl(
            presence=presence or gate,
            x=sx,
            y=sy,
            z=hand.z_norm,
            pan=0.0,
            pitch_norm=pitch_norm,
            mod=0.0,
            gesture=gesture if gesture != "None" else None,
            gate_open=gate,
            scale_degree=degree,
        )

    def _map_hand(
        self,
        hand: SmoothedHandFeatures,
        role: HandRole,
        gesture: str,
        pitch_hand: bool,
    ) -> HandControl:
        pitch_norm = _axis_value(hand, role.pitch_axis)
        sx = getattr(hand, "spatial_x", hand.x_norm)
        sy = getattr(hand, "spatial_y", hand.y_norm)
        return HandControl(
            presence=hand.presence,
            x=sx,
            y=sy,
            z=hand.z_norm,
            pan=0.0,
            pitch_norm=pitch_norm,
            mod=0.0,
            gesture=gesture if gesture != "None" else None,
            gate_open=False,
            scale_degree=None,
        )

    def _resolve_pitch_degree(self, pitch_norm: float) -> int:
        n = scale_size(self._scale_name)
        max_deg = max(0, n - 1)
        release_high = 0.5 - EDGE if max_deg > 1 else 1.0 - EDGE

        if pitch_norm <= EDGE:
            self._edge_lock = "low"
        elif self._edge_lock == "low" and pitch_norm > RELEASE_LOW:
            self._edge_lock = None

        if pitch_norm >= 1.0 - EDGE:
            self._edge_lock = "high"
        elif self._edge_lock == "high" and pitch_norm < release_high:
            self._edge_lock = None

        if self._edge_lock == "low":
            return 0
        if self._edge_lock == "high":
            return max_deg

        return self._pitch_quantizer.quantize(pitch_norm)

    def gesture_action(self, gesture: str, hand: str | None) -> dict | None:
        trigger = self._config.gesture_triggers.get(gesture)
        if not trigger:
            return None
        th = str(trigger.get("hand", "any")).lower()
        if th != "any" and hand and th != hand.lower():
            return None
        return trigger
